core/src/sumeragi/fault.rs

```diff
@@ -38,866 +37,1337 @@ impl FaultInjection for NoFault {
-                .clone()
-                .into_builder()
-                .with_peers(wsv_peers)
-                .build()
-                // TODO: Check it during instruction execution.
-                .expect("The safety of changing the number of peers should have been checked at Instruction execution stage.");
-        }
-    }
-
-    /// Returns `true` if some block is in discussion, `false` otherwise.
-    #[allow(clippy::unused_async)]
-    pub async fn voting_in_progress(&self) -> bool {
-        self.voting_block.is_some()
-    }
-
-    /// Latest block hash as seen by sumeragi.
-    pub fn latest_block_hash(&self) -> &HashOf<VersionedCommittedBlock> {
-        self.topology.at_block()
-    }
-
-    /// Number of view changes.
-    /// Where a view change is a change in topology made if there was some consensus misfunction during round (faulty peers).
-    pub fn number_of_view_changes(&self) -> u64 {
-        self.topology.view_change_proofs().len() as u64
-    }
-
-    /// The proofs of view changes that happened after the last block was committed.
-    pub fn view_change_proofs(&self) -> &ViewChangeProofs {
-        self.topology.view_change_proofs()
-    }
-
-    /// The hash of the latest view change.
-    pub fn latest_view_change_hash(&self) -> HashOf<Proof> {
-        self.view_change_proofs().latest_hash()
-    }
+#[allow(clippy::expect_used)]
+fn request_view_change<F>(
+    sumeragi: &SumeragiWithFault<F>,
+    state_machine: &mut SumeragiStateMachineData,
+    view_change_proof_chain: &mut Vec<Proof>,
+    current_view_change_index: u64,
+) where
+    F: FaultInjection,
+{
+    let suspect_proof = {
+        let mut proof = Proof {
+            latest_block_hash: state_machine.latest_block_hash,
+            view_change_index: current_view_change_index,
+            signatures: Vec::new(),
+        };
+        proof
+            .sign(sumeragi.key_pair.clone())
+            .expect("Suspect proof must be able to perform signing");
+        proof
+    };
+
+    let _ = view_change_proof_chain.insert_proof(
+        &state_machine
+            .current_topology
+            .sorted_peers()
+            .iter()
+            .cloned()
+            .collect(),
+        state_machine.current_topology.max_faults(),
+        &state_machine.latest_block_hash,
+        &suspect_proof,
+    );
+
+    sumeragi.broadcast_packet(
+        MessagePacket::new(
+            view_change_proof_chain.clone(),
+            Message::ViewChangeSuggested,
+        ),
+        &state_machine.current_topology,
+    );
+}
 
-    /// Get peers as a hash set of their ids.
-    pub fn peers(&self) -> HashSet<PeerId> {
-        self.topology.sorted_peers().iter().cloned().collect()
-    }
+#[instrument(skip(sumeragi, state_machine_guard))]
+#[allow(clippy::expect_used)]
+/// Execute the main loop of [`SumeragiWithFault`]
+pub fn run_sumeragi_main_loop<F>(
+    sumeragi: &SumeragiWithFault<F>,
+    mut state_machine_guard: SumeragiStateMachineData,
```

Comment:
Not guard anymore.

Reply:
Yea, I didn't rename it yet because I was making other changes.